components/service/pocket/src/main/java/mozilla/components/service/pocket/PocketEndpoint.kt

```diff
@@ -0,0 +1,75 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.pocket
+
+import android.support.annotation.WorkerThread
+import mozilla.components.concept.fetch.Client
+import mozilla.components.service.pocket.data.PocketGlobalVideoRecommendation
+import mozilla.components.service.pocket.net.PocketResponse
+
+// See @Ignore tests in test class to do end-to-end testing.
+
+/**
+ * Makes requests to the Pocket API and returns the requested data.
+ *
+ * @see [getInstance] to retrieve an instance.
+ */
+class PocketEndpoint internal constructor(
+    private val rawEndpoint: PocketEndpointRaw,
+    private val jsonParser: PocketJSONParser
+) {
+
+    /**
+     * Gets a response, filled with the Pocket global video recommendations from the Pocket API server on success.
+     * This network call is synchronous and the results are not cached. The API version of this call is available in the
+     * source code at [PocketURLs.VALUE_VIDEO_RECS_VERSION].
+     *
+     * If the API returns unexpectedly formatted results, these entries will be omitted and the rest of the items are
+     * returned.
+     *
+     * @return a [PocketResponse.Success] with the Pocket global video recommendations (the list will never be empty)
+     * or, on error, a [PocketResponse.Failure].
+     */
+    @WorkerThread // Synchronous network call.
+    fun getGlobalVideoRecommendations(): PocketResponse<List<PocketGlobalVideoRecommendation>> {
+        val json = rawEndpoint.getGlobalVideoRecommendations()
+        val videoRecs = json?.let { jsonParser.jsonToGlobalVideoRecommendations(it) }
+        return videoRecs.wrapInResponse()
+    }
+
+    private fun <T> List<T>?.wrapInResponse(): PocketResponse<List<T>> = if (isNullOrEmpty()) {
+        PocketResponse.Failure()
+    } else {
```

Comment:
I wonder if `Failure` should wrap an exception. It would allow the caller to log more meaningful errors. It would also allows you to not map everything to "null" and then to failure and instead just wrap the exception?

Reply:
That's a good question. I programmed it like I was working on a local application (log it ASAP and convert it into an easier error type to deal with than Exception) rather than a public API but it might make sense to expose this data in the API.

That being said, propagating exceptions adds more complexity than just returning an error value like null (i.e. as an API owner, I should catch all known exceptions and convert them into predictable types) and, as a consumer, I doubt I'd do anything other than log it (i.e. I wouldn't handle different errors differently) so I don't think it's worth that complexity right now. We can always add a `cause` property to `Failure` without introducing a breaking change later so I'd vote we stick with what we have now and add the Exception if consumers ask for it.

@pocmo What do you think?

Reply:
> (i.e. as an API owner, I should catch all known exceptions and convert them into predictable types)

But isn't that what you are doing already? 

> We can always add a cause property to Failure without introducing a breaking change later so I'd vote we stick with what we have now and add the Exception if consumers ask for it.

Alright!

Reply:
> > (i.e. as an API owner, I should catch all known exceptions and convert them into predictable types)
> 
> But isn't that what you are doing already?

😰